---
title: "scRNA QC"
output: html_document
date: "`r Sys.Date()`"
params:
  ## If you have Ribosomal ratio in your raw seurat object put this as TRUE otherwise leave as FALSE
  ribosomal: FALSE
  params_file: https://github.com/bcbio/bcbioR-test-data/raw/refs/heads/main/singlecell/parameters.R
  project_file: ../information.R
---

```{r, cache = FALSE, message = FALSE, warning=FALSE}
# This set up the working directory to this file so all files can be found
library(rstudioapi)
setwd(fs::path_dir(getSourceEditorContext()$path))
# NOTE: This code will check version, this is our recommendation, it may work
#.      other versions
stopifnot(R.version$major>= 4) # requires R4
if (compareVersion(R.version$minor,"3.1")<0) warning("We recommend >= R4.3.1") 
stopifnot(compareVersion(as.character(BiocManager::version()), "3.18")>=0)
stopifnot(compareVersion(as.character(packageVersion("Seurat")), "5.0.0")>=0)

source(params$params_file)
source(params$project_file)
```

This code is in this ![](https://img.shields.io/badge/status-alpha-yellow) revision.

```{r, eval=FALSE}
### READ ME FIRST

# This is a template for scRNA QC to present to your client. The actual QC can be done using our rshiny app: 

# After you have decided on your QC metrics load your raw object (i.e. right after you first read data into seurat) and put the parameters.R file you got from the shiny app in the same folder as this rmd.

```

# Overview

-   Project: `r project`
-   PI: `r PI`
-   Analyst: `r analyst`
-   Experiment: `r experiment`

```{r setup, include=FALSE}
library(Seurat)
library(tidyverse)
library(ggplot2)

# 1. Set up input files in this R file (params_de.R)
# Loading QC'd object
parameters = params$params_file
if (isUrl(seurat_obj)){
  source(url(parameters))
}else{
  source(parameters)
}

seurat_obj = "~/Downloads/filtered_gene_bc_matrices/tiny.rds"

knitr::opts_chunk[["set"]](
    cache = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE,
    fig.height = 4,
    echo=FALSE)
```



```{r load }
## Load data
seurat_raw <- readRDS(seurat_obj)
```


```{r find doublets, eval = F}
# if you suspect that your dataset contains doublets, you can use this code to 
# detect them and filter them out. if your seurat object contains data from more
# than one sample, it is important to pass the metadata column containing the 
# sample name to scDblFinder using the "samples" argument

sce <- as.SingleCellExperiment(seurat_raw, assay = 'RNA')
sce <- scDblFinder(sce, samples = "orig.ident")

meta_scdblfinder <- sce@colData@listData %>%
    as.data.frame() %>%
    dplyr::select(starts_with("scDblFinder")) %>%
  mutate(barcode = sce@colData@rownames)

seurat_raw_meta <- seurat_raw@meta.data

seurat_raw_meta_new <- seurat_raw_meta %>% left_join(meta_scdblfinder, by = 'barcode')
rownames(seurat_raw_meta_new) <- seurat_raw_meta_new$barcode
seurat_raw@meta.data <- seurat_raw_meta_new

seurat_raw<- subset(seurat_raw, scDblFinder.class == 'singlet')


```

```{r filter no ribo, eval = !ribosomal, warning=FALSE, results='asis'}
## Creat QC object USE METRICS YOU CHOSE IN THE RSHINY APP

seurat_qc <- subset(x = seurat_raw,
                    subset = (nCount_RNA >= nCount_RNA_cutoff)
                    & (nFeature_RNA >= nFeature_RNA_cutoff) 
                    & (mitoRatio < mitoRatio_cutoff)
                ##  & (riboRatio < riboRatio_cutoff)  
                    & (Log10GenesPerUMI > Log10GenesPerUMI_cutoff)
                    )


```


```{r filter ribo, eval=ribosomal, warning=FALSE, results='asis'}

seurat_qc <- subset(x = seurat_raw,
                    subset = (nCount_RNA >= nCount_RNA_cutoff)
                    & (nFeature_RNA >= nFeature_RNA_cutoff) 
                    & (mitoRatio < mitoRatio_cutoff)
                    & (riboRatio < riboRatio_cutoff)  
                    & (Log10GenesPerUMI > Log10GenesPerUMI_cutoff)
                    )

```


```{r}

## Save QC object
saveRDS(seurat_qc, file = "seurat_post-QC.rds")
```


```{r prep-info}
## Prep information for plotting 
metadata0 <- seurat_raw@meta.data
metadata0 = metadata0 %>%  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
metadata1 <- seurat_qc@meta.data
metadata1 = metadata1 %>%  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
```


# QC metrics: raw data {.tabset}

In this section, we review quality control (QC) metrics for the **raw feature matrices** generated by `Cellranger`. Only a low level filter excluding cells with <100 nUMIs (= number of unique molecular identifiers, or sequenced reads per cell) was applied when uploading the data into `R`. 


## Cells per sample

```{r cells raw}
table(metadata0$orig.ident)

```


## UMIs per cell

Here, we look at the distribution of UMIs (unique molecular identifiers, or sequenced reads) per cell (droplet) in the dataset. Before QC, we expect a biomodal distribution with a first *small* peak at low numbers of UMIs (<250) corresponding to droplets that encapsulated background/dying cells, and a second higher peak centered at >1000. The line is at 250.


```{r raw_nUMIs}
metadata0 %>% 
  	ggplot(aes(x = nUMI, color = orig.ident, fill =  orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() + 
  	ylab("Cell density") + scale_x_log10() +
  	geom_vline(xintercept = 250) +
    facet_wrap(. ~ orig.ident) +
    ggtitle("UMIs per cell in raw dataset")
```



```{r}
# Visualize the distribution of nUMIs per cell (boxplot)
metadata0 %>% 
  	ggplot(aes(x=orig.ident, y=log10(nUMI), fill=orig.ident)) + 
  	geom_violin() + geom_boxplot(width = 0.1, fill = alpha("white", 0.7)) +
  	theme_classic() +
    geom_hline(yintercept = c(log10(1000), log10(50000))) +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  	theme(plot.title = element_text(hjust = 0.5, face = "bold")) 
```

## Genes per cell

Here, we look at the number of different genes that were detected in each cell. By "detected", we mean genes with a non-zero read count measurement. Gene detection in the range of 500 to 5000 is normal for most single-cell experiments. The line is at 750. 

```{r raw_nGene}
# Visualize the distribution of genes detected per cell (histogram)
metadata0 %>% 
  	ggplot(aes(x = nGene, color = orig.ident, fill = orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = c(700)) +
    facet_wrap(. ~ orig.ident) +
    ggtitle("Detected genes per cell in raw dataset")
```


```{r}
# Visualize the distribution of nUMIs per cell (boxplot)
metadata0 %>% 
  	ggplot(aes(x=orig.ident, y=log10(nGene), fill=orig.ident)) + 
  	geom_violin() + geom_boxplot(width = 0.1, fill = alpha("white", 0.7)) +
  	theme_classic() +
    geom_hline(yintercept = c(log10(700))) +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  	theme(plot.title = element_text(hjust = 0.5, face = "bold")) 
```


## Mitochondrial ratio

We evaluate overall mitochondrial gene expression as a biomarker of cellular stress during sample preparation. Typically, we expect mitochondrial genes to account for <20% of overall transcripts in each cell. The line indicates 10%.

```{r raw_mito, warning=FALSE}
# Visualize the distribution of mitochondrial gene expression detected per cell
metadata0 %>% 
  	ggplot(aes(color = orig.ident, x = mitoRatio, fill = orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = c(0.1)) + 
    facet_wrap(. ~ orig.ident) +
    ggtitle("Percentage of mitochondrial gene expression per cell in raw dataset")
```



```{r raw_ribo, eval=ribosomal, warning=FALSE, results='asis'}

cat("## Ribosomal ratio \n")

cat("We evaluate overall ribosomal gene expression. The line indicates 5%. \n"
)
# Visualize the distribution of mitochondrial gene expression detected per cell
metadata0 %>% 
  	ggplot(aes(color = orig.ident, x = riboRatio, fill = orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = c(0.05)) + 
    facet_wrap(. ~ orig.ident) +
    ggtitle("Percentage of ribosomal gene expression per cell in raw dataset")
```


## UMIs vs. Genes

By plotting the number of UMIs per cell (x-axis) vs. the number of genes per cell (y-axis), we can visually assess whether there is a large proportion of low quality cells with low read counts and/or gene detection (bottom left quadrant of the plot). In the following representation, cells are further color-coded based on the percentage of mitochondrial genes found among total detected genes. The line for nUMI is at 1000 and the line for nGene is at 700.

```{r raw_gene_by_umi, fig.height=12, fig.width=15, warning=FALSE}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata0 %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 1000) +
  	geom_hline(yintercept = 700) + 
    ggtitle("Genes vs. nUMIs in raw dataset") + 
  	facet_wrap(~orig.ident)
```


## Complexity

Another way to assess the quality and purity of a single-cell dataset is to look for cells that have fewer detected genes per UMI than others. Typical values for this metric are >0.8 for most cells. Cells with lower diversity in the genes they express may be low-complexity cell types such as red blood cells. With sorted populations, we expect high purity and a very similar complexity distribution across samples.

```{r raw_novelty}
# Visualize the overall novelty of the gene expression by visualizing the genes detected per UMI
metadata0 %>%
  	ggplot(aes(x = Log10GenesPerUMI, color = orig.ident, fill = orig.ident)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = c(0.85)) +
    facet_wrap(. ~ orig.ident) +
    ggtitle("log10(Genes per UMI) in raw dataset")
```


```{r}
# Visualize the distribution of nUMIs per cell (boxplot)
metadata0 %>% 
  	ggplot(aes(x=orig.ident, Log10GenesPerUMI, fill=orig.ident)) + 
  	geom_violin() + geom_boxplot(width = 0.1, fill = alpha("white", 0.7)) +
  	theme_classic() +
    geom_hline(yintercept = c(0.8, 0.85)) +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  	theme(plot.title = element_text(hjust = 0.5, face = "bold")) 
```


# QC metrics: Filtered data {.tabset}

Based on the above QC metrics, we filtered the dataset to isolate cells passing the following thresholds: >`nCount_RNA_cutoff` UMIs, >`nFeature_RNA_cutoff` genes, <`mitoRatio_cutoff` mitochondrial gene ratio, and >`Log10GenesPerUMI_cutoff` complexity.

In this section, we review QC metrics for our filtered dataset.

## Cells per sample

```{r cells filtered}
table(metadata1$orig.ident)

```


## UMIs per cell

The line is at 1000

```{r qc1_nUMIs}
metadata1 %>% 
  	ggplot(aes(color = orig.ident, x = nUMI, fill = orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") + xlab("nUMI") +
    geom_vline(xintercept = c(1000)) +
    facet_wrap(. ~ orig.ident)
```


```{r}
# Visualize the distribution of nUMIs per cell (boxplot)
metadata1 %>% 
  	ggplot(aes(x=orig.ident, y=log10(nUMI), fill=orig.ident)) + 
  	geom_violin() + geom_boxplot(width = 0.1, fill = alpha("white", 0.7)) +
  	theme_classic() +
    geom_hline(yintercept = c(log10(1000))) +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  	theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```


## Genes detected

The line is at 750

```{r qc1_genes}
# Visualize the distribution of genes detected per cell via histogram
metadata1 %>% 
  	ggplot(aes(color = orig.ident, x = nGene, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + xlab("nGene") +
    facet_wrap(. ~ orig.ident) +
  	geom_vline(xintercept = c(750))
```


```{r}
# Visualize the distribution of nUMIs per cell (boxplot)
metadata1 %>% 
  	ggplot(aes(x=orig.ident, y=log10(nGene), fill=orig.ident)) + 
  	geom_violin() + geom_boxplot(width = 0.1, fill = alpha("white", 0.7)) +
  	theme_classic() +
    geom_hline(yintercept = c(log10(7500))) +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  	theme(plot.title = element_text(hjust = 0.5, face = "bold")) 

```


## Mitochondrial ratio

The line is at 10%.

```{r qc1_mitoratio, message=FALSE, warning=FALSE}
# Visualize the distribution of mitochondrial gene expression detected per cell
metadata1 %>% 
  	ggplot(aes(color = orig.ident, x = mitoRatio, fill = orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
    geom_vline(xintercept = 0.1) +
    facet_wrap(. ~ surgery)
```


```{r qc1_ribo, eval=ribosomal, warning=FALSE, results='asis'}

cat("## Ribosomal ratio \n")

cat("We evaluate overall ribosomal gene expression. The line indicates 10%. \n"
)
# Visualize the distribution of mitochondrial gene expression detected per cell
metadata0 %>% 
  	ggplot(aes(color = orig.ident, x = riboRatio, fill = orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = c(0.1)) + 
    facet_wrap(. ~ orig.ident) +
    ggtitle("Percentage of ribosomal gene expression per cell in raw dataset")
```


## UMIs vs. Genes

Both the horizontal and vertical lines are at 1000.

```{r qc1_genes_per_UMI, fig.height=12, fig.width=15, warning=FALSE}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata1 %>% 
  	ggplot(aes(x = nUMI, y = nGene, color = mitoRatio)) + 
  	geom_point() + 
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = c(1000)) +
  	geom_hline(yintercept = c(1000)) + 
    ggtitle("Genes vs. nUMIs in raw dataset") + 
    xlab("nUMI") + ylab("nGene") + 
  	facet_wrap(~orig.ident)
```


## Complexity

```{r qc1_complexity}
# Visualize the overall novelty of the gene expression by visualizing the genes detected per UMI
metadata1 %>%
  	ggplot(aes(x = Log10GenesPerUMI, color = orig.ident, fill = orig.ident)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	#geom_vline(xintercept = c(0.85)) +
    facet_wrap(. ~ orig.ident)
```

```{r}
# Visualize the distribution of nUMIs per cell (boxplot)
metadata1 %>% 
  	ggplot(aes(x=orig.ident, Log10GenesPerUMI, fill=orig.ident)) + 
  	geom_violin() + geom_boxplot(width = 0.1, fill = alpha("white", 0.7)) +
  	theme_classic() +
    #geom_hline(yintercept = c(0.85)) +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  	theme(plot.title = element_text(hjust = 0.5, face = "bold")) 
```


# R session

```{r}
sessionInfo()
```
